File: Striver A2Z/Step 13 - Binary Trees [Traversals, Medium and Hard Problems]/Lec 1 - Traversals/binary-tree-level-order-traversal.cpp

// https://leetcode.com/problems/binary-tree-level-order-traversal/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(!root) return ans;

        queue<TreeNode*> q;
        q.push(root);
        q.push(NULL);
        vector<int> currentLevel;

        while(!q.empty()){
            auto front = q.front();
            q.pop();

            if(!front){
                if(q.empty()) break;
                ans.push_back(currentLevel);
                currentLevel.clear();
                q.push(NULL);
                continue;
            }

            currentLevel.push_back(front->val);
            if(front->left) q.push(front->left);
            if(front->right) q.push(front->right);
        }

        ans.push_back(currentLevel);
        return ans;
    }
};

File: Striver A2Z/Step 13 - Binary Trees [Traversals, Medium and Hard Problems]/Lec 1 - Traversals/binary-tree-level-order-traversal.js
// https://leetcode.com/problems/binary-tree-level-order-traversal/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  const ans = [];
  if (!root) return ans;

  const q = [root, null];
  let smallAns = [];
  while (!(q.length === 0)) {
    let front = q.shift();

    if (!front) {
      if (q.length === 0) break;
      ans.push(smallAns);
      smallAns = [];
      q.push(null);
      continue;
    }

    smallAns.push(front.val);
    if (front.left) q.push(front.left);
    if (front.right) q.push(front.right);
  }

  ans.push(smallAns);
  return ans;
};
